### PR TITLE
TY: Finally dismantled `Ty.unifyWith` & lot of other type inference changes

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 1]
```diff
@@ -45,54 +44,76 @@ private val RsTraitItem.typeParamSingle: TyTypeParameter?
...
     }
+    private val intoIteratorTraitAndOutput: Pair<RsTraitItem, RsTypeAlias>? by lazy(NONE) {
+        val trait = items.findCoreItem("iter::IntoIterator") as? RsTraitItem ?: return@lazy null
+        findAssociatedType(trait, "Item")?.let { trait to it }
+    }
+
+    private val ctx: RsInferenceContext
+        get() {
+            if (_ctx == null)
+                _ctx = RsInferenceContext()
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: Braces or `ctrl+shilf+j` or `get() = _ctx ?: RsInferenceContext().apply { _ctx = this }`

File: src/main/kotlin/org/rust/lang/core/types/infer/Fulfillment.kt [hunk 2]
```diff
@@ -98,6 +98,8 @@ class ObligationForest {
         nodes.add(Node(obligation))
     }
 
+    var printed = false
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/types/infer/Fulfillment.kt]: debug print

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 3]
```diff
@@ -432,88 +523,123 @@ private class RsFnInferenceContext(
...
-                .map { TraitRef(ty, it) }
-                .map { it.foldTyTypeParameterWith { map[it] ?: it } }
-                .map(this::normalizeAssociatedTypesIn)
-                .forEach { fulfill.registerPredicateObligation(Obligation(Predicate.Trait(it))) }
+        val map = run {
+            val map = typeParameters + element.generics.associate { it to ctx.typeVarForParam(it) }
+            if (selfTy != null) {
+                val selfTySubst = selfTy.substitute(map)
+                map.substituteInValues(mapOf(TyTypeParameter.self() to selfTySubst)) +
+                    mapOf(TyTypeParameter.self() to selfTySubst)
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: I think `+ (key to value)` should work as well.

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 4]
```diff
@@ -45,54 +44,76 @@ private val RsTraitItem.typeParamSingle: TyTypeParameter?
...
+            return _ctx!!
+        }
 
-    fun findImplsAndTraits(ty: Ty): Collection<BoundElement<RsTraitOrImpl>> {
-        if (ty is TyTypeParameter) return ty.getTraitBoundsTransitively()
-        return findImplsAndTraitsCache.getOrPut(project, ty) { rawFindImplsAndTraits(ty) }
+    fun findImplsAndTraits(ty: Ty): Collection<RsTraitOrImpl> {
+        if (ty is TyTypeParameter) return ty.getTraitBoundsTransitively().map { it.element }
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: We receive BoundElements from path resolution, so we can't just skip it. What we can do in this place - put this line under the cache. Now it's possible

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 5]
```diff
@@ -217,83 +320,117 @@ class ImplLookup(private val project: Project, private val items: StdKnownItems)
...
-        val (derefTrait, derefTarget) = derefTraitAndTarget ?: return null
-        val (_, subst) = findImplOfTrait(ty, derefTrait) ?: return null
-        return subst[derefTarget]
+    fun findIteratorItemType(ty: Ty): TyWithObligations<Ty>? {
+        return selectProjection(iteratorTraitAndOutput ?: return legacyFindIteratorItemType(ty), ty).ok()
+            ?: selectProjection(intoIteratorTraitAndOutput ?: return null, ty).ok()
     }
 
-    fun findIteratorItemType(ty: Ty): Ty {
+    // TODO legacy; used in tests only; should be removed
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: Hmm, it's private and the name prefixed with "legacy". It's enough, I guess...

**Pull request merged**